cpp/src/arrow/dataset/partition.cc

```diff
@@ -561,6 +636,74 @@ class DirectoryPartitioningFactory : public KeyValuePartitioningFactory {
   std::vector<std::string> field_names_;
 };
 
+class FilenamePartitioningFactory : public KeyValuePartitioningFactory {
+ public:
+  FilenamePartitioningFactory(std::vector<std::string> field_names,
+                              PartitioningFactoryOptions options)
+      : KeyValuePartitioningFactory(options), field_names_(std::move(field_names)) {
+    Reset();
+    util::InitializeUTF8();
+  }
+
+  std::string type_name() const override { return "filename"; }
+
+  Result<std::shared_ptr<Schema>> Inspect(
+      const std::vector<std::string>& paths) override {
+    for (const auto& path : paths) {
+      size_t field_index = 0;
```

Comment:
Why not make this an `int`?

Reply:
We compare this with the size of `field_names_`, so defining this as a `size_t` should be appropriate?

Reply:
Oh, I see this is essentially copy-pasted from `DirectoryPartitioningFactory::Inspect`. Is it possible to factor this out to avoid significant repetitions of code?